### PR TITLE
Xclbin to ELF flow migration

### PR DESCRIPTION
**Pull request by rbramand-xilinx**
#### Problem solved by the commit
This PR enables new flow where ELF file is input instead of xclbin.
spec - https://confluence.amd.com/display/AIE/AIE+Compiler+Artifacts#AIECompilerArtifacts-Flow2

spec of new config elf - https://confluence.amd.com/pages/viewpage.action?pageId=1485002156#Profile/Config/CtrlcodeElfSpec-Config.elf

This new Elf  has partition size and kernel signature.
It has multiple .ctrltext and .ctrldata sections and multiple .pdi sections.
Each ctrltext section represents one control code and pdi addresses along with kernel args needs to be patched into the control code buffer.
Control code that needs to be run is decided while creating xrt::kernel object using kernel name  -  <kernel_name>:<ctrl_code_id>   eg: "DPU:0"  runs with control code from section .ctrltext0  and  "DPU:1" runs with control code from section .ctrlcode1

Sample test case :
```
#include "xrt/xrt_hw_context.h"
#include "xrt/xrt_kernel.h"
#include "xrt/xrt_elf.h"

int main(int argc, char** argv)
{
    std::string elf_path {argv[1]};   // elf path
    auto elf = xrt::elf(elf_path);

    auto device = xrt::device(0);
    auto ctx = xrt::hw_context(device, elf);

    std::string kernel_name = "DPU:1";
    auto kernel = xrt::ext::kernel(ctx, kernel_name);
    
    //  create args
    auto run = kernel(arg1, arg2 ...);
    run.wait();

    return 0;
}
```

XRT first class object changes :
xrt::hw_context  -  Added new APIs to create context using Elf instead of xclbin
xrt::kernel  -  Added new ext API to create kernel object using kernel name, added changes to construct kernel args from kernel signature.
xrt::elf  -  can now take new Elf with os abi 70 as input
xrt::module  -  added changes to parse this new Elf, also refactored code

**TODO** :  Refactor code to parse ELF file using xrt::elf class and make it similar to xrt::xclbin. Created CR - https://jira.xilinx.com/browse/CR-1219757 for tracking this.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
This is a new feature

#### How problem was solved, alternative solutions (if any) and why they were rejected
Hw context can be created either with xclbin (traditional flow) or with xrt::elf. A new constructor is provided for same.
A new xrt::ext::kernel constructor is provided that takes ctx and kernel name as input.
Added code to parse new Elf file and patch it according to spec.
Rest of the flow remains same

#### Risks (if any) associated the changes in the commit
Low to medium
Tested the existing flows but needs more testing with all the available test cases.

#### What has been tested and how, request additional testing if necessary
Tested with new application flow on aie2p simnow  (needs changes with respect to flow in amdxdna shim and firmware, changes are yet to be merged, so tested with local drops)

Tested existing test cases on phoenix hw (linux) and tests passes so existing flow didn't break.

TODO : check whether existing aie2ps test cases work

#### Documentation impact (if any)
Added doxygen comments in code for new APIs added, may be we need to document about new flow after it is stabilized.

### CODE REVIEW

**Review by stsoe — CHANGES_REQUESTED** [review]
Overall looks good.  I am not completely done reviewing, but maybe address my first points and I will review again.

**Comment by AShivangi** [comment]
@hlaccabu recently added some code to enable elf flow in xrt-smi validate for npu3. Please work with him to make sure that this change doesn't breaks the existing code.

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 1]
```diff
@@ -1383,16 +1628,22 @@ class module_sram : public module_impl
...
+      return fill_ert_aie2ps(payload);
+    else if (os_abi == Elf_Amd_Aie2p_config)
+      return fill_ert_aie2p_preempt_data(payload);
 
-    return fill_ert_aie2ps(payload);
+    // os abi is Elf_Amd_Aie2p
+    if (m_preempt_save_bo && m_preempt_restore_bo)
+      return fill_ert_aie2p_preempt_data(payload);
+    else
+      return fill_ert_aie2p_non_preempt_data(payload);
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:1640]: Maybe a switch/case is better than if/else here since we have 3+ osabi

File: src/runtime_src/core/include/ert.h [hunk 2]
```diff
@@ -626,28 +626,29 @@ struct cu_cmd_state_timestamps {
...
+  ERT_CLK_CALIB            = 13,
+  ERT_MB_VALIDATE          = 14,
+  ERT_START_KEY_VAL        = 15,
+  ERT_ACCESS_TEST_C        = 16,
+  ERT_ACCESS_TEST          = 17,
+  ERT_START_DPU            = 18,
+  ERT_CMD_CHAIN            = 19,
+  ERT_START_NPU            = 20,
+  ERT_START_NPU_PREEMPT    = 21,
+  ERT_START_NPU_PDI_IN_ELF = 22,
```
Comment (larry9523) [inline src/runtime_src/core/include/ert.h:651]: Let's start with a different name: ERT_START_NPU_PREEMPT_ELF. (With this flow, PDI is not necessarily in ELF.)

File: src/runtime_src/core/common/api/hw_context_int.h [hunk 3]
```diff
@@ -43,6 +44,12 @@ XRT_CORE_COMMON_EXPORT
 xrt::hw_context
 create_hw_context_from_implementation(void* hwctx_impl);
 
+xrt::xclbin::kernel
+get_kernel(const xrt::hw_context& hwctx);
+
+xrt::module
+get_module(const xrt::hw_context& hwctx, const std::string& kname);
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/hw_context_int.h]: Hi Soren, get_kernel function was mistakenly added and I forgot to remove it.
Also I will document the functions which are newly added. Thanks

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 4]
```diff
@@ -53,6 +59,25 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+    , m_mode{mode}
+  {}
+
+  hw_context_impl(std::shared_ptr<xrt_core::device> device, const xrt::elf& elf, cfg_param_type cfg_param, access_mode mode)
+    : m_core_device{std::move(device)}
+    , m_cfg_param{std::move(cfg_param)}
+    , m_mode{mode}
+  {
+    auto module = xrt::module(elf);
+    auto kernel_name = xrt_core::module_int::get_kernel_name(module);
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp:74]: All the Elf parsing is done in xrt_module.cpp and this was our design from the beginning even before this change.
So whenever we create xrt::module object Elf is parsed and info is stored in module object for future retrieval.
Yes as per spec there will be only one kernel per Elf but there can be multiple control codes.
I will add more comments in the code where ever possible.

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 5]
```diff
@@ -81,6 +106,33 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+      m_partition_size = part_size;
+      m_hdl = m_core_device->create_hw_context(m_partition_size, m_cfg_param, m_mode);
+      return;
+    }
+
+    // add module to map if kernel name is different, else throw
+    if (m_partition_size != part_size)
+      throw std::runtime_error("can not add config to ctx with different configuration\n");
+
+    for (const auto& m : m_module_map) {
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp:130]: Yes Soren you are right, I will modify the code

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 6]
```diff
@@ -81,6 +106,33 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
   hw_context_impl& operator=(const hw_context_impl&) = delete;
   hw_context_impl& operator=(hw_context_impl&&) = delete;
 
+  void
+  add_config(const xrt::elf& elf)
+  {
+    auto module = xrt::module(elf);
+    auto kernel_name = xrt_core::module_int::get_kernel_name(module);
+    auto part_size = xrt_core::module_int::get_partition_size(module);
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp:114]: All the Elf parsing is done using module object Soren so info can be retrieved using it.

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 7]
```diff
@@ -129,6 +181,16 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
   {
     return m_usage_logger.get();
   }
+
+  xrt::module
+  get_module(const std::string& kname) const
+  {
+    for (const auto& m : m_module_map) {
+      if (kname == xrt_core::module_int::get_kernel_name(m.second))
+        return m.second;
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp:191]: Yes Soren that simplifies things, Thanks. I will modify the code

File: src/runtime_src/core/include/ert.h (hunk 2 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/include/ert.h:651]: sure Larry

**Comment by rbramand-xilinx** [comment]
> good to go from driver perspective. As long as driver's xrt submodule (and corresponding xrt_plugin.deb) contains these changes as well as the XRT .deb.

@hlaccabu can you please check if your feature doesn't break with these changes

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 8]
```diff
@@ -1520,6 +1521,64 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
...
+    std::stringstream ss(s);
+    std::string item;
+
+    while (getline(ss, item, delimiter))
+      tokens.push_back(item);
+
+    return tokens;
+  }
+
+  void
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1537]: cant make it static Soren as it uses member variables (args vector)

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 9]
```diff
@@ -1520,6 +1521,64 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
...
+  construct_elf_kernel_args(const std::string& kernel_name)
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1538]: Yes Soren, I can move this functionality to xrt_module.cpp as that file is used for elf parsing

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 10]
```diff
@@ -1306,7 +1306,7 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
   xrt::xclbin::kernel xkernel;         // kernel xclbin metadata
   std::vector<argument> args;          // kernel args sorted by argument index
   std::vector<ipctx> ipctxs;           // CU context locks
-  const property_type& properties;     // Kernel properties from XML meta
+  property_type properties;            // Kernel properties from XML meta
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1309]: sure Soren thanks for the suggestions, I will modify the code accordingly

**Comment by rbramand-xilinx** [comment]
Hi @stsoe , created CR - https://jira.xilinx.com/browse/CR-1219757 for refactoring existing code to move ELF parsing to xrt::elf object and make it similar to xrt::xclbin.
Please review this PR, I will do the refactoring using next set of PRs

File: src/runtime_src/core/common/api/module_int.h [hunk 11]
```diff
@@ -54,6 +57,19 @@ get_ert_opcode(const xrt::module& module);
...
+
+std::string
+get_kernel_signature(const xrt::module& module);
+
+std::string
+get_kernel_name(const xrt::module& module);
+
+// Get partition size if ELF has info
+uint32_t
+get_partition_size(const xrt::module& module);
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/module_int.h]: done

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 12]
```diff
@@ -53,6 +59,25 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+    : m_core_device{std::move(device)}
+    , m_cfg_param{std::move(cfg_param)}
+    , m_mode{mode}
+  {
+    auto module = xrt::module(elf);
+    auto kernel_name = xrt_core::module_int::get_kernel_name(module);
+    m_module_map[kernel_name] = std::move(module);
+
+    m_partition_size = xrt_core::module_int::get_partition_size(m_module_map.begin()->second);
+    m_hdl = m_core_device->create_hw_context(m_partition_size, m_cfg_param, mode);
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp]: done

**Comment by larry9523** [comment]
@rbramand-xilinx , Did we run any exsiting preemption test to make sure we won't break any?

File: src/runtime_src/core/common/api/module_int.h [hunk 13]
```diff
@@ -33,7 +40,8 @@ patch(const xrt::module&, const std::string& argnm, size_t index, const xrt::bo&
 // of the control code buffer.
 XRT_CORE_COMMON_EXPORT
 void
-patch(const xrt::module&, uint8_t*, size_t*, const std::vector<std::pair<std::string, uint64_t>>*);
+patch(const xrt::module&, uint8_t*, size_t*, const std::vector<std::pair<std::string, uint64_t>>*,
```
Comment (stsoe) [inline src/runtime_src/core/common/api/module_int.h:43]: It would be good to document the arguments, in particular the new `index` argument.

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 14]
```diff
@@ -1,13 +1,17 @@
...
+// Copyright (C) 2022-2024 Advanced Micro Devices, Inc. All rights reserved.
 
 // This file implements XRT xclbin APIs as declared in
 // core/include/experimental/xrt_queue.h
 #define XRT_API_SOURCE         // exporting xrt_hwcontext.h
 #define XCL_DRIVER_DLL_EXPORT  // exporting xrt_xclbin.h
 #define XRT_CORE_COMMON_SOURCE // in same dll as coreutil
+
+#include "core/include/experimental/xrt_module.h"
 #include "core/include/xrt/xrt_hw_context.h"
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:11]: swap line 10 and 11, first include normally is the header for the cpp.

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 15]
```diff
@@ -81,6 +109,32 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
 
+  void
+  add_config(const xrt::elf& elf)
+  {
+    auto module = xrt::module(elf);
+    auto kernel_name = xrt_core::module_int::get_kernel_info(module).props.name;
+    auto part_size = xrt_core::module_int::get_partition_size(module);
+
+    // create hw ctx handle if not already created
+    if (m_hdl == nullptr) {
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:120]:
```suggestion
    if (!m_hdl) {
```

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 16]
```diff
@@ -81,6 +109,32 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+      m_module_map[kernel_name] = std::move(module);
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:121]:
```suggestion
      m_module_map.emplace(std::move(kernel_name), std::move(module));
```

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 17]
```diff
@@ -81,6 +109,32 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+    }
+
+    // add module only if partition size matches existing configuration
+    if (m_partition_size != part_size)
+      throw std::runtime_error("can not add config to ctx with different configuration\n");
+
+    // add module to map if kernel name is different, else throw
+    if (m_module_map.find(kernel_name) != m_module_map.end())
+      throw std::runtime_error("config with kernel already exists, cannot add this config\n");
+    m_module_map[kernel_name] = std::move(module);
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:135]:
```suggestion
    m_module_map.emplace(std::move(kernel_name), std::move(module));
```

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 18]
```diff
@@ -129,6 +183,15 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
     return m_usage_logger.get();
   }
+
+  xrt::module
+  get_module(const std::string& kname) const
+  {
+    auto itr = m_module_map.find(kname);
+    if (itr == m_module_map.end())
+      throw std::runtime_error("no module found with given kernel name in ctx");
+    return itr->second;
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:193]:
```suggestion
    if (auto itr = m_module_map.find(kname); itr != m_module_map.end())
       return itr->second;
       
    throw std::runtime_error("no module found with given kernel name in ctx");
```

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 19]
```diff
@@ -207,6 +276,41 @@ alloc_hwctx_from_mode(const xrt::device& device, const xrt::uuid& xclbin_id, xrt
...
+  // Update device is called with a raw pointer to dyanamically
+  // link to callbacks that exist in XDP via a C-style interface
+  // The create_hw_context_from_implementation function is then 
+  // called in XDP create a hw_context to the underlying implementation
+  xrt_core::xdp::update_device(handle.get());
+
+  handle->get_usage_logger()->log_hw_ctx_info(handle.get());
+
+  return handle;
+}
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:312]: Little concerning with duplicated code.  How about?


```suggestion
std::shared_ptr<hw_context_impl>
static post_allow_hwctx(const std::shared_ptr<hw_context_impl>& handle)
{
  // Update device is called with a raw pointer to dyanamically
  // link to callbacks that exist in XDP via a C-style interface
  // The create_hw_context_from_implementation function is then 
  // called in XDP create a hw_context to the underlying implementation
  xrt_core::xdp::update_device(handle.get());

  handle->get_usage_logger()->log_hw_ctx_info(handle.get());

  return handle;
}

static std::shared_ptr<hw_context_impl>
alloc_empty_hwctx(const xrt::device& device, const xrt::hw_context::cfg_param_type& cfg_param, xrt::hw_context::access_mode mode)
{
  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
  return post_alloc_hwctx(std::make_shared<hw_context_impl>(device.get_handle(), cfg_param, mode));
}


static std::shared_ptr<hw_context_impl>
alloc_hwctx_from_elf(const xrt::device& device, const xrt::elf& elf, const xrt::hw_context::cfg_param_type& cfg_param,
                     xrt::hw_context::access_mode mode)
{
  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
  return post_alloc_hwctx(std::make_shared<hw_context_impl>(device.get_handle(), elf, cfg_param, mode));
}
```

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 20]
```diff
@@ -1086,7 +1086,7 @@ class argument
     size_t size;   // size in bytes of argument per xclbin
 
     explicit
-    global_type(size_t bytes)
+    global_type(size_t bytes = 0)
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_kernel.cpp:1089]: Is this change even used?

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 21]
```diff
@@ -1316,6 +1316,7 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
   size_t num_cumasks = 1;              // Required number of command cu masks
   control_type protocol = control_type::none; // Default opcode
   uint32_t uid;                        // Internal unique id for debug
+  uint32_t m_ctrl_code_index = 0;      // Index to identify which ctrl code to load in elf
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_kernel.cpp:1319]: Thanks for using `m_`, I have long regretted not having started out that way for all the other members.

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 22]
```diff
@@ -1504,6 +1505,20 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
...
 
+  static xrt::module
+  get_module(const xrt::hw_context& ctx, const std::string& kname)
+  {
+    // ELF use case, identify module from ctx that has given kernel name
+    // kernel name will be of format - <kernel_name>:<index>
+    auto i = kname.find(":");
+    if (i == std::string::npos) {
+      // default case
+      return xrt_core::hw_context_int::get_module(ctx, kname.substr(0, kname.size()));
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_kernel.cpp:1516]: Isn't this just `kname`?  As in 
```suggestion
      return xrt_core::hw_context_int::get_module(ctx, kname);
```

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 23]
```diff
@@ -1579,10 +1594,36 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
...
+    auto i = nm.find(":");
+    if (i == std::string::npos) {
+      // default case - ctrl code 0 will be used
+      name = nm.substr(0, nm.size());
+      m_ctrl_code_index = 0;
+    }
+    else {
+      name = nm.substr(0, i);
+      m_ctrl_code_index = std::stoul(nm.substr(i+1, nm.size()-i-1));
+    }
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_kernel.cpp:1617]: Uhh, this is a little hard to parse.
This is the second place name is parsed at `:`.  Can this parsing be re-factored into say
```
std::pair<std::string, uint32_t>
static get_kernel_name_and_ctrl_code_index(const std::string& nm)
{
   ...
   return {name, index};
}
```
Then use as 
```
std::tie(name, m_ctrl_code_index) = get_name_and_ctrl_code_index(nm);
```
Maybe choose a better function name for the refactored code.

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 24]
```diff
@@ -988,6 +955,38 @@ get_kernel_arguments(const axlf* top, const std::string& kname)
...
+// Kernel auto restart counter offset
+// Needed until meta-data support (Vitis-1147)
+kernel_properties::restart_type
+get_restart_from_ini(const std::string& kname)
+{
+  static auto restart_kernels = xrt_core::config::get_auto_restart_kernels();
+  return (restart_kernels.find("/" + kname + "/") != std::string::npos)
+    ? 1
+    : 0;
+}
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/xclbin_parser.cpp:980]: Do we need to export these functions?

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 25]
```diff
@@ -988,6 +955,38 @@ get_kernel_arguments(const axlf* top, const std::string& kname)
...
+
+// Kernel software reset
+bool
+get_sw_reset_from_ini(const std::string& kname)
+{
+  static auto reset_kernels = xrt_core::config::get_sw_reset_kernels();
+  return (reset_kernels.find("/" + kname + "/") != std::string::npos);
+}
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/xclbin_parser.cpp:988]: Same export question

File: src/runtime_src/core/common/xclbin_parser.h [hunk 26]
```diff
@@ -424,6 +424,15 @@ get_project_name(const axlf* top);
...
 
+kernel_properties::mailbox_type
+get_mailbox_from_ini(const std::string& kname);
+
+kernel_properties::restart_type
+get_restart_from_ini(const std::string& kname);
+
+bool
+get_sw_reset_from_ini(const std::string& kname);
+
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/xclbin_parser.h:435]: Still open?

File: src/runtime_src/core/include/experimental/xrt_module.h [hunk 27]
```diff
@@ -85,14 +85,19 @@ class module : public detail::pimpl<module_impl>
...
    * with the specified hardware context.
+   * If module has multiple control codes, index is used to identify
+   * the control code that needs to be run.
+   * By default control code at zeroth index is picked
    *
    * Throws if module is not compatible with hardware context
    */
   XRT_API_EXPORT
-  module(const xrt::module& parent, const xrt::hw_context& hwctx);
+  module(const xrt::module& parent, const xrt::hw_context& hwctx, uint32_t ctrl_code_idx = 0);
```
Comment (stsoe) [inline src/runtime_src/core/include/experimental/xrt_module.h:100]: This is a breaking change.   Existing code will no longer work because the function with no argument is no longer present.  Avoid default arguments, just create a new ctor.

**Review by stsoe — CHANGES_REQUESTED** [review]
Some more comments.  I did not review xrt_module.cpp.

File: src/runtime_src/core/common/xclbin_parser.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/xclbin_parser.h:435]: I need these functions while creating properties in ELF flow as well @stsoe , should I use static functions there as well? I moved these functions in header so that there wont be code duplicity

File: src/runtime_src/core/common/api/xrt_kernel.cpp (hunk 21 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1319]: Thanks Soren

File: src/runtime_src/core/common/api/xrt_hw_context.cpp (hunk 19 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_hw_context.cpp:312]: Yes Soren, this removes code duplicity. Thanks

File: src/runtime_src/core/include/experimental/xrt_module.h (hunk 27 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/include/experimental/xrt_module.h:100]: As I was using default arg existing application should work right Soren.
Yeah I will use new constructor instead of default arg.

File: src/runtime_src/core/common/api/xrt_kernel.cpp (hunk 20 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1089]: actually for global_type size is not used but it was present before my changes so made it use default arg.
May be I will remove the arg completely as its not used.

File: src/runtime_src/core/common/api/xrt_kernel.cpp (hunk 23 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp:1617]: I will refactor the code Soren, my attempt was to use initializer list for most of the class members but yeah it does not look good.

File: src/runtime_src/core/include/experimental/xrt_module.h (hunk 27 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/include/experimental/xrt_module.h]: Yes understood Soren, changed the code to have 2 constructors

File: src/runtime_src/core/common/api/xrt_kernel.cpp (hunk 20 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp]: I have removed the size arg completely as it is not used in any member function of the class

Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_kernel.cpp]: Huh?  The size argument is used at line 1185.   Can you just leave this global_type ctor unchanged?

Comment (rbramand-xilinx) [reply src/runtime_src/core/common/api/xrt_kernel.cpp]: sure Soren, I will add it back.
But the size is used only for construction in 1185 but not used in the class constructed is what I was telling.

File: src/runtime_src/core/common/api/xrt_hw_context.cpp [hunk 28]
```diff
@@ -81,6 +109,31 @@ class hw_context_impl : public std::enable_shared_from_this<hw_context_impl>
...
+    }
+
+    // add module only if partition size matches existing configuration
+    if (m_partition_size != part_size)
+      throw std::runtime_error("can not add config to ctx with different configuration\n");
+
+    // add module to map if kernel name is different, else throw
+    if (m_module_map.find(kernel_name) != m_module_map.end())
+      throw std::runtime_error("config with kernel already exists, cannot add this config\n");
+    m_module_map.emplace(std::move(kernel_name), std::move(module));
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_hw_context.cpp:134]:
```suggestion

    m_module_map.emplace(std::move(kernel_name), std::move(module));
```

File: src/runtime_src/core/common/api/xrt_kernel.cpp [hunk 29]
```diff
@@ -1504,6 +1505,15 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
     return data;  // no skipping
   }
 
+  static uint32_t
+  get_ctrlcode_idx(const std::string& name)
+  {
+    // kernel name will be of format - <kernel_name>:<ctrl code index>
+    if (auto i = name.find(":"); i != std::string::npos)
+      return std::stoul(name.substr(i+1, name.size()-i-1));
+    return 0; // default case
```
Comment (stsoe) [inline src/runtime_src/core/common/api/xrt_kernel.cpp:1514]:
```suggestion

    return 0; // default case
```

**Review by stsoe — COMMENTED** [review]
Looks good now. Let me know when you have resolved if we really need to export internal xclbin_parser functions or not.

**Comment by rbramand-xilinx** [comment]
> Looks good now. Let me know when you have resolved if we really need to export internal xclbin_parser functions or not.

Hi @stsoe, we don't need those functions to be exported. I have reverted the change and verified the testcase it passes.
Thanks for the insights.

**Review by stsoe — APPROVED** [review]

**Comment by rbramand-xilinx** [comment]
Rebased with master.
Testing existing flows with rebased version so adding do not merge tag till testing is done.

**Review by larry9523 — APPROVED** [review]
Should we move the copyright to 2025?

**Comment by rbramand-xilinx** [comment]
> Should we move the copyright to 2025?

Updated, Thanks Larry.

**Review by chvamshi-xilinx — APPROVED** [review]
